posix/poll.c

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <openenclave/enclave.h>
+
+#include <openenclave/corelibc/sys/poll.h>
+#include <openenclave/internal/posix/fdtable.h>
+#include <openenclave/internal/posix/raise.h>
+#include "posix_t.h"
+
+int oe_poll(struct oe_pollfd* fds, oe_nfds_t nfds, int timeout)
+{
+    int ret = -1;
+    int retval = -1;
+    struct oe_host_pollfd* host_fds = NULL;
+    oe_nfds_t i;
+
+    if (!fds || nfds == 0)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    if (!(host_fds = oe_calloc(nfds, sizeof(struct oe_host_pollfd))))
+        OE_RAISE_ERRNO(OE_ENOMEM);
+
+    /* Convert enclave fds to host fds. */
+    for (i = 0; i < nfds; i++)
+    {
+        oe_host_fd_t host_fd;
+        oe_fd_t* desc;
+
+        /* Fetch the fd struct for this fd struct. */
+        if (!(desc = oe_fdtable_get(fds[i].fd, OE_FD_TYPE_ANY)))
```

Comment:
negative file descriptors are actually supported and documented by `poll`:
```
       The field fd contains a file descriptor for an open file.  If this
       field is negative, then the corresponding events field is ignored and
       the revents field returns zero.  (This provides an easy way of ignor‐
       ing a file descriptor for a single poll() call: simply negate the fd
       field.  Note, however, that this technique can't be used to ignore
       file descriptor 0.)
```
I would only do the fd table + host fd lookup if `fds[i].fd > 0`.